pkg/cmd/server/kubernetes/master/master_config.go

```diff
@@ -166,67 +190,132 @@ func BuildDefaultAPIServer(options configapi.MasterConfig) (*apiserveroptions.Se
 	storageFactory.AddCohabitatingResources(extensions.Resource("horizontalpodautoscalers"), autoscaling.Resource("horizontalpodautoscalers"))
 
-	return server, storageFactory, nil
+	return storageFactory, nil
 }
 
-// TODO this function's parameters need to be refactored
-func BuildKubernetesMasterConfig(
-	options configapi.MasterConfig,
-	requestContextMapper apirequest.RequestContextMapper,
-	kubeClient kclientset.Interface,
-	informers shared.InformerFactory,
-	admissionControl admission.Interface,
-	originAuthenticator authenticator.Request,
-	kubeAuthorizer authorizer.Authorizer,
-) (*MasterConfig, error) {
-	if options.KubernetesMasterConfig == nil {
-		return nil, errors.New("insufficient information to build KubernetesMasterConfig")
+// buildUpstreamGenericConfig copies the apiserver.Config setup code from k8s.io/kubernetes/cmd/kube-apiserver/app/server.go.
+// ONLY COMMENT OUT CODE HERE, do not modify it. Do modifications outside of this function.
+func buildUpstreamGenericConfig(s *kapiserveroptions.ServerRunOptions) (*apiserver.Config, error) {
+	// set defaults
+	if err := s.GenericServerRunOptions.DefaultAdvertiseAddress(s.SecureServing, s.InsecureServing); err != nil {
+		return nil, err
+	}
+	// In origin: certs should be available:
+	//_, apiServerServiceIP, err := master.DefaultServiceIPRange(s.ServiceClusterIPRange)
+	//if err != nil {
+	//	return nil, fmt.Errorf("error determining service IP ranges: %v", err)
+	//}
+	//if err := s.SecureServing.MaybeDefaultWithSelfSignedCerts(s.GenericServerRunOptions.AdvertiseAddress.String(), apiServerServiceIP); err != nil {
+	//	return nil, fmt.Errorf("error creating self-signed certificates: %v", err)
+	//}
+	if err := s.CloudProvider.DefaultExternalHost(s.GenericServerRunOptions); err != nil {
+		return nil, fmt.Errorf("error setting the external host value: %v", err)
 	}
 
-	// in-order list of plug-ins that should intercept admission decisions
-	// TODO: Push node environment support to upstream in future
+	s.Authentication.ApplyAuthorization(s.Authorization)
 
-	podEvictionTimeout, err := time.ParseDuration(options.KubernetesMasterConfig.PodEvictionTimeout)
+	// validate options
+	if errs := s.Validate(); len(errs) != 0 {
+		return nil, kerrors.NewAggregate(errs)
+	}
+
+	// create config from options
+	genericConfig := apiserver.NewConfig().
+		WithSerializer(kapi.Codecs)
+
+	if err := s.GenericServerRunOptions.ApplyTo(genericConfig); err != nil {
+		return nil, err
+	}
+	if err := s.Etcd.ApplyTo(genericConfig); err != nil {
+		return nil, err
+	}
+	if err := s.SecureServing.ApplyTo(genericConfig); err != nil {
+		return nil, err
+	}
+	if err := s.InsecureServing.ApplyTo(genericConfig); err != nil {
+		return nil, err
+	}
+	if err := s.Audit.ApplyTo(genericConfig); err != nil {
+		return nil, err
+	}
+	if err := s.Features.ApplyTo(genericConfig); err != nil {
+		return nil, err
+	}
+	if err := s.Authentication.ApplyTo(genericConfig); err != nil {
```

Comment:
This is only about client CA. Also compare https://github.com/sttts/origin/pull/88/commits/ef440b4e05453c4bc0214afc6a5e5c96f02bc653, the request header case is done here.